snuba/consumer.py

```diff
@@ -82,5 +104,66 @@ def flush_batch(self, batch):
 
             self.producer.flush()
 
+    @staticmethod
+    def validate_message(message):
+        """
+        Validate a raw mesage from kafka, with 1 of 3 results.
+          - A tuple of (action_type, data) is returned to be handled by the proper
+            handler/processor for that action type.
+          - None is returned, indicating we should silently ignore this message.
+          - An exeption is raised, inditcating something is wrong and we should stop.
+        """
+        action_type = None
+        data = None
+
+        if isinstance(message, dict):
+            # deprecated unwrapped event message == insert
+            action_type = INSERT
+            data = message
+        elif isinstance(message, (list, tuple)) and len(message) >= 2:
+            version = message[0]
+
+            if version in (0, 1, 2):
+                # version 0: (0, 'insert', data)
+                # version 1: (1, type, data, [state])
+                #   NOTE: types 'delete_groups', 'merge' and 'unmerge' are ignored
+                # version 2: (2, type, data, [state])
+                type_, event = message[1:3]
+                if type_ == 'insert':
+                    action_type = INSERT
+                    data = event
+                else:
+                    if version == 0:
+                        raise InvalidMessageType("Invalid message type: {}".format(type_))
+                    elif version == 1:
+                        if type_ in ('delete_groups', 'merge', 'unmerge'):
```

Comment:
The consumer/validation changes made sense to me until about here: these types are `events` dataset specific.

And so is the protocol version, really, since it is versioned when we make changes to the `events` messages. I'm trying to understand how session stuff will fit into this?

Reply:
yeah good question. I think the protocol here is sufficient for both events and spans. at least the `insert` action is common to both, unfortunately yeah, the replacement actions are all individual actions, rather than being grouped under a `replace` action so I ended up checking them at this level to be consistent with the insert.

I'm not sure what the best plan is, have the dataset/processor expose a list of `REPLACE_ACTIONS` or something that indicate valid actions that we would throw on the replacements topic?

Reply:
It's a bit weird because technically the protocol expects/implies this for `insert` (using the current version, 2, as an example):

```
extra_data=({
            'group_id': event.group_id,
            'event_id': event.event_id,
            'organization_id': project.organization_id,
            'project_id': event.project_id,
            # TODO(mitsuhiko): We do not want to send this incorrect
            # message but this is what snuba needs at the moment.
            'message': event.message,
            'platform': event.platform,
            'datetime': event.datetime,
            'data': event_data,
            'primary_hash': primary_hash,
            'retention_days': retention_days,
        }, {
            'is_new': is_new,
            'is_sample': is_sample,
            'is_regression': is_regression,
            'is_new_group_environment': is_new_group_environment,
            'skip_consume': skip_consume,
        },)
```

It's *two* dicts with those expected keys. At least that's what it has been so far. If we removed one of those keys or something we'd bump to version 3 which seems unrelated to session data? I feel like protocol version is tightly coupled to the actions and their expected bodies.

But maybe it's possible we just ignore that and datasets care (or don't) about certain actions and the data (which could be 1 or more objects, like `insert`) are blindly passed down the dataset to do whatever?

Reply:
Yeah, hmm, so I guess the fully consistent way would be for each dataset to define its own kafka message protocol. the flipside being that we will end up with a bunch of redefinitions of the simple commands/verbs like "insert" that are going to be common to everything.

Hmmm, this looks like it will be kinda messy whichever way we do it.